File: extension/src/dsl.ts
// Detect a Max DSL command in editor text. Mirrors the engine grammar:
//   [sigil] <operator> body <operator>
// sigil ∈ { @ # ! } (optional), operator ∈ { . , .. , ~ }.
// The raw matched text is sent verbatim to /command — the engine re-parses the
// sigil + operator and routes accordingly.

export type Operator = "generate" | "summarize" | "fix";

export interface DslMatch {
  text: string; // the trimmed command, sent as-is to the engine
  operator: Operator;
  sigil: string; // "" | "@" | "#" | "!"
  body: string;
  cloud: boolean; // sigil === "!"
}

const SUMMARIZE = /^([@#!]?)\.\.\s?([\s\S]*?)\s?\.\.$/;
const FIX = /^([@#!]?)~\s?([\s\S]*?)\s?~$/;
const GENERATE = /^([@#!]?)\.\s?([\s\S]*?)\s?\.$/;

// Try `..` before `.` (the single-dot pattern would otherwise match a `..` line).
const RULES: [Operator, RegExp][] = [
  ["summarize", SUMMARIZE],
  ["fix", FIX],
  ["generate", GENERATE],
];

export function detectCommand(raw: string): DslMatch | null {
  const text = raw.trim();
  if (!text) return null;
  for (const [operator, re] of RULES) {
    const m = re.exec(text);
    // Require real content (a word char), so `..  ..` / `.  .` aren't commands.
    if (m && /\w/.test(m[2])) {
      const sigil = m[1] ?? "";
      return { text, operator, sigil, body: m[2].trim(), cloud: sigil === "!" };
    }
  }
  return null;
}

File: clients/nvim/lua/max/dsl.lua
-- Detect a Max DSL command in editor text. Mirrors the engine grammar:
--   [sigil] <operator> body <operator>
-- sigil in { @ # ! % ^ } (optional), operator in { . , .. , ~ }.
-- The raw matched text is sent verbatim to /command — the engine re-parses the
-- sigil + operator and routes accordingly.
--
-- Cloud sigils ("!" Claude, "%" OpenAI) leave the machine; "@" "#" "^" stay local.

local M = {}

-- Lua patterns are not full regex (no non-greedy quantifiers), so we anchor on
-- the literal delimiters and capture the middle greedily, then trim. Order
-- matters: try ".." before "." so a ".. .." line isn't matched as ". .".
-- Captures: (1) optional sigil, (2) body.
local RULES = {
  { op = "summarize", pat = "^([@#!%%%^]?)%.%.(.*)%.%.$" },
  { op = "fix", pat = "^([@#!%%%^]?)~(.*)~$" },
  { op = "generate", pat = "^([@#!%%%^]?)%.(.*)%.$" },
}

local CLOUD = { ["!"] = true, ["%"] = true }

-- Trim leading/trailing whitespace.
local function trim(s)
  return (s:gsub("^%s+", ""):gsub("%s+$", ""))
end

--- Detect a command in `raw`. Returns a table or nil.
--- @return table|nil  { text, operator, sigil, body, cloud }
function M.detect_command(raw)
  local text = trim(raw or "")
  if text == "" then
    return nil
  end
  for _, rule in ipairs(RULES) do
    local sigil, body = text:match(rule.pat)
    if body ~= nil then
      body = trim(body)
      -- Require real content (a word char) so ".  ." / "..  .." aren't commands.
      if body:match("%w") then
        sigil = sigil or ""
        return {
          text = text,
          operator = rule.op,
          sigil = sigil,
          body = body,
          cloud = CLOUD[sigil] == true,
        }
      end
    end
  end
  return nil
end

return M
